src/solver/mha/mha_solver_backward.cpp

```diff
@@ -0,0 +1,475 @@
+
+namespace miopen {
+
+namespace solver {
+
+namespace mha {
+
+namespace { // TODO: Issue #2748
+template <typename T, typename S = std::enable_if_t<std::is_unsigned_v<T>, T>>
+constexpr inline S Ceil(const T val, const T div)
+{
+    return (val - 1 + div) / div;
+}
+
+template <typename T, typename S = std::enable_if_t<std::is_unsigned_v<T>, T>>
+constexpr S RoundUpToMultiple(T val, T mul)
+{
+    return Ceil(val, mul) * mul;
+}
+
+constexpr uint32_t nextPow2(uint32_t v)
+{
+    if(v == 1)
+    {
+        return (v << 1);
+    }
+    else
+    {
+        v--;
+        v |= v >> 1;
+        v |= v >> 2;
+        v |= v >> 4;
+        v |= v >> 8;
+        v |= v >> 16;
+        v++;
+        return v;
+    }
+}
+
+MultiBufferWorkspaceTraits SplitBufferToWorkspace(size_t S, size_t D, size_t NHS)
+{
+    // the first MatMuls (N*H*S*D) * (N*H*S*D)T = (N*H*S*S)
+    // the second MatMuls (N*H*S*S)[T] * (N*H*S*D) = (N*H*S*D)
+    // dOxO row reduction (N*H*S*1)
+    return MultiBufferWorkspaceTraits{NHS * S * get_data_size(miopenFloat), // first matmuls
+                                      NHS * S * get_data_size(miopenFloat), // first matmuls
+                                      NHS * std::max(S, D) *
+                                          get_data_size(miopenFloat), // reduction and second matmul
+                                      NHS * D * get_data_size(miopenFloat),  // second matmuls
+                                      NHS * D * get_data_size(miopenFloat)}; // second matmuls
+}
+
+MultiBufferWorkspaceTraits SplitBufferToWorkspace(const std::vector<size_t>& lengths)
+{
+    const auto [N, H, S, D] = miopen::tien<4>(lengths);
+    return SplitBufferToWorkspace(S, D, N * H * S);
+}
+
+miopen::HipEventPtr make_hip_fast_event()
+{
+    hipEvent_t result = nullptr;
+    hipEventCreateWithFlags(&result, hipEventDisableTiming);
+    return miopen::HipEventPtr{result};
+}
+} // namespace
+
+bool MhaBackward::IsApplicable([[maybe_unused]] const ExecutionContext& context,
+                               const miopen::mha::ProblemDescription& problem) const
+{
+    const miopen::mha::MhaInputDescsBackward& descsForward = problem.GetDescsBackward();
+
+    auto [N, H, S, D] = miopen::tien<4>(descsForward.kDesc.GetLengths());
+
+    return !miopen::IsDisabled(ENV(MIOPEN_DEBUG_ATTN_NAIVE_BWD)) && //
+           !problem.IsForward() &&                                  //
+           S <= std::numeric_limits<uint32_t>::max() &&             //
+           D <= std::numeric_limits<uint32_t>::max() &&             //
+           descsForward.kDesc.IsPacked() &&                         //
+           descsForward.qDesc.IsPacked() &&                         //
+           descsForward.vDesc.IsPacked() &&                         //
+           descsForward.oDesc.IsPacked() &&                         //
+           descsForward.doDesc.IsPacked() &&                        //
+           MIOPEN_USE_GEMM;
```

Comment:
[Recommendation][Readability] Usually I move operators to the next line, like this:
```C++
    return !miopen::IsDisabled(ENV(MIOPEN_DEBUG_ATTN_NAIVE_BWD)) //
           && !problem.IsForward() &&                            //
           ...
           && descsForward.doDesc.IsPacked()                     //
           && MIOPEN_USE_GEMM;
```

Reply:
As prefixes, ok.